File: src/ast/impl.cpp
#include "impl.h"
#include "ast.h"
#include "function.h"
#include "mod.h"
#include "scope.h"
#include "trait.h"
#include <algorithm>
#include <fmt/core.h>
#include <memory>
#include <vector>

namespace lygos {
    namespace AST {
        Impl::Impl(std::string type, Block body, std::vector<Type::Generic> generics, std::string trait):
            AST(ASTType::Impl), type(type), body(body), generics(generics), trait(trait) {

        }

        std::string Impl::GetValue() {
            return type;
        }

        static void ValidateTraitImpl(Impl *impl, Trait::Trait *trait, Scope *scope) {
            if(scope->GetStruct(impl->Type()).ImplementsTrait(trait->GetValue()))
                Log::Logger::Warn(fmt::format("type `{}` already implements `{}`", impl->Type(), trait->GetValue()));
            //TODO: find the excat function(s) that are not implemented and
            //check args/ret type
            /*(for(size_t i = 0; i < trait->Functions().Body().size(); i++) {
                bool found = false;
                for(size_t j = 0; j < impl->Body().size(); j++) {
                    if(((Function *)trait->Functions().Body()[i].get())->GetName() == ((Function *)impl->Body()[j].get())->GetName()) {
                        found = true;
                }
                if(!found)
                    Log::Logger::Warn("trait impl does not implement all functions");
                }
            }
            auto search = [](Ref<AST> fn) {return false;};
            if(std::find_if(impl->Body().begin(), impl->Body().end(), search) != impl->Body().end()) {

            }*/
            //std::mismatch();
        }

        llvm::Value *Impl::GenCode(Scope *scope) {
            auto &strct = scope->GetStruct(type);
            if(trait != "") {
                ValidateTraitImpl(this, &scope->GetTrait(trait), scope);
                strct.RegisterTraitImpl(trait);
            }
            for(const auto &member : body.Body()) {
                member->GenCode(scope);
            }
            return nullptr;
        }

        Ref<Type::Type> Impl::GetType(Scope *scope) { return nullptr; }

        void Impl::Lower(AST *parent) {
            if(parent->type == ASTType::MacroCall)
                return;
            auto &strct = ast_root->GetCurrentBlock()->Scope().GetStruct(type);
            for(const auto &gen : this->generics) {
                if(!strct.generics.contains(gen.name))
                    Log::Logger::Warn(fmt::format("differing generics for type `{}` in impl. `{}`", strct.name, gen.name));
            }

            for(size_t i = 0; i < body.Body().size(); i++) {
                Function *func = (Function*)body.Body()[i].get();
                std::string name_mangeled = {type + "_" + func->GetName()};
                strct.AddFunction({func->GetName(), name_mangeled, func->GetArgs(), func->GetRetType(), func->IsMember()});
                func->GetName() = name_mangeled;
                func->GetMangledName() = name_mangeled;
            }

            LYGOS_ASSERT(parent->type == ASTType::Mod);
            for(size_t i = 0; i < body.Body().size(); i++) {
                ast_root->SetCurrentBlock(&((Mod *)parent)->Body());
                body.Body()[i]->Lower(parent);
                body.Increment();
            }
        }

        void Impl::Sanatize() {

        }
    }
}

File: src/ast/impl.rs
use crate::types::Type;
use crate::lexer::Tagged;
use crate::ast::{Block, Generate, AST};
use crate::ast::symbol;

#[derive(Debug)]
pub struct Impl {
    pub typ: Tagged<String>,
    pub body: Block,
    pub trat: Option<Tagged<String>>,
}

impl Impl {
    pub fn new(typ: Tagged<String>, body: Block, trat: Option<Tagged<String>>) -> Self {
        Self{ typ, body, trat }
    }
}

impl Generate for Impl {
    fn loc(&self) -> &crate::lexer::Loc {
        self.typ.loc()
    }

    fn get_value(&self) -> String {
        self.typ.inner().to_string()
    }

    fn gen_code(&mut self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<llvm::ValueRef> {
        for func in &mut self.body.body {
            if let AST::Function(func) = func {
                func.gen_code(scope, ctx);
            }
        }
        None
    }

    fn get_type(&self, _: &mut super::Scope, _: &crate::GenerationContext) -> Option<Type> { None }

    fn collect_symbols(&mut self, scope: &mut super::Scope) {
        for func in &mut self.body.body {
            if let AST::Function(func) = func {
                let strct = if let Some(strct) = scope.try_resolve_symbol(&self.typ) {
                    if let symbol::Symbol::Struct(strct) = strct {
                        strct
                    }else {
                        scope.add_symbol(self.typ.inner(), symbol::Symbol::Struct(symbol::Struct::new_dummy(self.typ.clone())));
                        scope.get_struct(&self.typ)
                    }
                }else {
                    scope.add_symbol(self.typ.inner(), symbol::Symbol::Struct(symbol::Struct::new_dummy(self.typ.clone())));
                    scope.get_struct(&self.typ)
                };

                let mut func_sym: symbol::Function = func.into();
                func_sym.name_mangeled = strct.name.inner().to_owned() + "_" + func_sym.name.inner();
                strct.register_function(func_sym);
            }
        }
        if let Some(t) = &self.trat {
            let strct = scope.get_struct(&self.typ);
            strct.register_trait_impl(t);
        }
    }
}
